File: lib/sbyc/r/system/operators/domain.rb
SByC::R.operators.define{

  operator{|op|
    op.description = %Q{ Returns name of a domain }
    op.signature   = s(system::Domain)
    op.argnames    = [:operand]
    op.returns     = system::String
    op.aliases     = [:'name-of']
  }
  def name_of(operand) operand.domain_name; end

  operator{|op|
    op.description = %Q{ Checks if a domain is a subdomain of another one }
    op.signature   = s(system::Domain, system::Domain)
    op.argnames    = [:'sub-domain', :domain]
    op.returns     = system::Boolean
    op.aliases     = [:'is-sub-domain-of?']
  }
  def is_sub_domain_of?(subdomain, domain)
    subdomain.is_sub_domain_of?(domain)
  end
  
  operator{|op|
    op.description = %Q{ Checks if a domain is a proper subdomain of another one }
    op.signature   = s(system::Domain, system::Domain)
    op.argnames    = [:'sub-domain', :domain]
    op.returns     = system::Boolean
    op.aliases     = [:'is-proper-sub-domain-of?']
  }
  def is_proper_sub_domain_of?(subdomain, domain)
    subdomain.is_proper_sub_domain_of?(domain)
  end
  
  operator{|op|
    op.description = %Q{ Checks if a domain is a super domain of another one }
    op.signature   = s(system::Domain, system::Domain)
    op.argnames    = [:'super-domain', :domain]
    op.returns     = system::Boolean
    op.aliases     = [:'is-super-domain-of?']
  }
  def is_super_domain_of?(superdomain, domain)
    is_sub_domain_of?(domain, superdomain)
  end
  
  operator{|op|
    op.description = %Q{ Checks if a domain is a proper super domain of another one }
    op.signature   = s(system::Domain, system::Domain)
    op.argnames    = [:'super-domain', :domain]
    op.returns     = system::Boolean
    op.aliases     = [:'is-proper-super-domain-of?']
  }
  def is_proper_super_domain_of?(superdomain, domain)
    is_proper_sub_domain_of?(domain, superdomain)
  end
  
}

File: test/spec/r/r_scripts/domain.r
# domain-of
(should (domain-of true)
        (eq Boolean))
(should (domain-of 12)
        (eq Integer))
        
# is-a?
(should (is-a? Boolean true)
        (eq true))
(should (is-a? Boolean "true")
        (eq false))
        
# is-sub-domain-of?
(should (is-sub-domain-of? Alpha Alpha)
        (eq true))
(should (is-sub-domain-of? Boolean Alpha)
        (eq true))
(should (is-sub-domain-of? Alpha Boolean)
        (eq false))
(should (is-sub-domain-of? Integer Numeric)
        (eq true))
(should (is-sub-domain-of? Integer Alpha)
        (eq true))
(should (is-sub-domain-of? Integer Float)
        (eq false))

# is-proper-sub-domain-of?
(should (is-proper-sub-domain-of? Boolean Alpha)
        (eq true))
(should (is-proper-sub-domain-of? Alpha Alpha)
        (eq false))
        
# is-super-domain-of?
(should (is-super-domain-of? Alpha Alpha)
        (eq true))
(should (is-super-domain-of? Alpha Boolean)
        (eq true))
(should (is-super-domain-of? Boolean Alpha)
        (eq false))
(should (is-super-domain-of? Numeric Integer)
        (eq true))
(should (is-super-domain-of? Alpha Integer)
        (eq true))
(should (is-super-domain-of? Float Integer)
        (eq false))

# is-proper-super-domain-of?
(should (is-proper-super-domain-of? Alpha Boolean)
        (eq true))
(should (is-proper-super-domain-of? Alpha Alpha)
        (eq false))
        
